reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -19,610 +19,1029 @@
+				return;
+			}
+
+			this.done = true;
+
+			final long previousState = markInboundTerminated(this);
+			if (hasInboundCancelled(previousState) || hasInboundTerminated(previousState)) {
+				return;
+			}
+
+			if (hasFirstValueSent(previousState)) {
+				this.inner.onComplete();
+				return;
+			}
+
+			if (!hasFirstValueReceived(previousState)) {
+				final Publisher<? extends R> result;
+				final CoreSubscriber<? super R> o = this.outer;
+
+				try {
+					result =
+							Objects.requireNonNull(this.transformer.apply(Signal.complete(
+									o.currentContext()), this),
+									"The transformer returned a null value");
+				}
+				catch (Throwable e) {
+					o.onError(e);
+					return;
+				}
+
+				result.subscribe(o);
+			}
+		}
+
+		@Override
+		public void cancel() {
+			long previousState = markInboundCancelled(this);
+			if (hasInboundCancelled(previousState) || hasInboundTerminated(previousState)) {
+				return;
+			}
+
+			this.s.cancel();
+
+			if (hasFirstValueReceived(previousState) && !hasInboundRequestedOnce(previousState)) {
+				final T f = this.firstValue;
+				Operators.onDiscard(f, currentContext());
+			}
+		}
+
+		void cancelAndError() {
+			long previousState = markInboundCancelled(this);
+			if (hasInboundCancelled(previousState) || hasInboundTerminated(previousState)) {
+				return;
+			}
+
+			this.s.cancel();
+
+			if (hasFirstValueReceived(previousState) && !hasInboundRequestedOnce(previousState)) {
+				final T f = this.firstValue;
+				Operators.onDiscard(f, currentContext());
+			}
+
+			if (hasInboundSubscriberSet(previousState)) {
+				this.inner.onError(new CancellationException("FluxSwitchOnFirst has already been cancelled"));
+			}
+		}
+
+		@Override
+		public void request(long n) {
+			if (Operators.validate(n)) {
+				// This is a sanity check to avoid extra volatile read in the request
+				// context
+				if (!this.isInboundRequestedOnce) {
+					this.isInboundRequestedOnce = true;
+
+					final T first = this.firstValue;
+					if (first != null) {
+						final long previousState = markInboundRequestedOnce(this);
+						if (hasInboundCancelled(previousState)) {
+							return;
+						}
+
+						final boolean wasDelivered = sendFirst(first);
+						if (wasDelivered) {
+							if (n != Long.MAX_VALUE) {
+								if (--n > 0) {
+									this.s.request(n);
+								}
+								return;
+							}
+						}
+					}
+				}
+
+				this.s.request(n);
+			}
+		}
+
+		boolean sendFirst(T firstValue) {
+			final CoreSubscriber<? super T> a = this.inner;
+
+			final boolean sent = tryOnNext(a, firstValue);
+
+			final long previousState = markFirstValueSent(this);
+			if (hasInboundCancelled(previousState)) {
+				return sent;
+			}
+
+			if (hasInboundTerminated(previousState)) {
+				Throwable t = this.throwable;
+				if (t != null) {
+					a.onError(t);
+				}
+				else {
+					a.onComplete();
+				}
+			}
+
+			return sent;
+		}
+
+		abstract boolean tryOnNext(CoreSubscriber<? super T> actual, T value);
```

Comment:
this is a bit confusing. `SwitchOnFirstMain` is not conditional, yet it has this method that _looks_ at a cursory glance like the one in `ConditionalSubscriber`. Then upon further inspection, it isn't the same, and in the case of `SwitchOnFirstConditionalMain` it doesn't even delegate to `tryOnNext(T value)` 🤔 

maybe rename it to something like `tryDirectSend`, as what it seems to be doing is bypassing all  the `tryOnNext` logic and passing down to the `actual`?